src/tv2-common/parts/server.ts

```diff
@@ -56,117 +58,245 @@ export function CreatePartServerBase<
+	const sanitisedScript = getScriptWithoutLineBreaks(partDefinition)
+	const actualDuration = getActualDuration(duration, sanitisedScript, props)
+
+	const displayTitle = getDisplayTitle(file, partDefinition)
+	const basePart = getBasePart(partDefinition, displayTitle, actualDuration, file, config.studio.CasparPrerollDuration)
+	const mediaPlayerSession = SanitizeString(`segment_${props.session ?? partDefinition.segmentExternalId}_${file}`)
+
+	const pieces: IBlueprintPiece[] = []
+
+	const serverBluePrintPiece = getServerBluePrintPiece(
+		partDefinition,
+		file,
+		actualDuration,
+		props,
+		layers,
+		sourceDuration,
+		mediaPlayerSession,
+		context,
+		config
+	)
+
+	const pgmBluePrintPiece = getPgmBluePrintPiece(
+		partDefinition,
+		file,
+		props,
+		layers,
+		sourceDuration,
+		mediaPlayerSession,
+		config
+	)
+
+	pieces.push(serverBluePrintPiece)
+	pieces.push(pgmBluePrintPiece)
+
+	return {
+		part: {
+			part: basePart,
+			adLibPieces: [],
+			pieces
+		},
+		file,
+		duration: actualDuration
 	}
+}
 
-	const file = partDefinition.fields.videoId
-	const mediaObjectDuration = context.hackGetMediaObjectDuration(file)
-	const sourceDuration =
-		mediaObjectDuration !== undefined ? mediaObjectDuration * 1000 - config.studio.ServerPostrollDuration : undefined
-	const duration =
+function isVideoIdMissing(partDefinition: PartDefinition): boolean {
+	return partDefinition.fields === undefined || !partDefinition.fields.videoId
+}
+
+function getVideoId(partDefinition: PartDefinition): string {
+	return partDefinition.fields.videoId ? partDefinition.fields.videoId : ''
+}
+
+function getSourceDuration(
+	mediaObjectDuration: number | undefined,
+	serverPostrollDuration: number
+): number | undefined {
+	return mediaObjectDuration !== undefined ? mediaObjectDuration * 1000 - serverPostrollDuration : undefined
+}
+
+function getDuration(
+	mediaObjectDuration: number | undefined,
+	sourceDuration: number | undefined,
+	props: ServerPartProps
+): number {
+	return (
 		(mediaObjectDuration !== undefined &&
 			((props.voLayer && props.totalWords <= 0) || !props.voLayer) &&
 			sourceDuration) ||
 		props.tapeTime * 1000 ||
 		0
-	const sanitisedScript = partDefinition.script.replace(/\n/g, '').replace(/\r/g, '')
-	const actualDuration =
-		props.voLayer && props.totalWords > 0
-			? (sanitisedScript.length / props.totalWords) * (props.totalTime * 1000 - duration) + duration
-			: duration
+	)
+}
 
-	const basePart = literal<IBlueprintPart>({
+function getActualDuration(duration: number, sanitisedScript: string, props: ServerPartProps): number {
+	return props.voLayer && props.totalWords > 0
+		? (sanitisedScript.length / props.totalWords) * (props.totalTime * 1000 - duration) + duration
+		: duration
+}
+
+function getDisplayTitle(file: string, partDefinition: PartDefinition): string {
+	if (partDefinition.type === PartType.VO || partDefinition.type === PartType.Server) {
+		return partDefinition.rawType
+	}
+	return file
```

Comment:
I mean line 154 should say `return 'SERVER'` because it's only reachable when the part is inserted dynamically as a result of queueing an adlib coming from an `ADLIBPIX=SERVER` cue that was in a part of a different type.